clang/tools/checked-c-convert/ArrayBoundsInferenceConsumer.cpp

```diff
@@ -1,127 +1,573 @@
+    }
+
+    return false;
+}
+
+static bool nameSubStringMatch(std::string ptrName, std::string lenFieldName) {
+  // Convert the names to lower case.
+  std::transform(ptrName.begin(), ptrName.end(), ptrName.begin(),
+                 [](unsigned char c){ return std::tolower(c); });
+  std::transform(lenFieldName.begin(), lenFieldName.end(), lenFieldName.begin(),
+                 [](unsigned char c){ return std::tolower(c); });
+  unsigned commonSubSeqLen = longestCommonSubsequence(ptrName.c_str(),
+                                                      lenFieldName.c_str(),
+                                                      ptrName.length(),
+                                                      lenFieldName.length());
+  if (commonSubSeqLen > 0) {
+    // Check if we get 80% match on the common subsequence matching on the
+    // variable name of length and the name of array.
+    return ((commonSubSeqLen * 100.0) / (ptrName.length() * 1.0)) >=
+           COMMONSUBSEQUENCEPERCMATCH;
+  }
+  return false;
+
+}
+
+static bool fieldNameMatch(std::string lenFieldName) {
+  // Convert the field name to lower case.
+  std::transform(lenFieldName.begin(), lenFieldName.end(), lenFieldName.begin(),
+                 [](unsigned char c){ return std::tolower(c); });
+  for (auto &potentialName : possibleLengthVarNamesPrefixes) {
+    if (lenFieldName.rfind(potentialName, 0) == 0)
+      return true;
+  }
+
+  for (auto &potentialName : possibleLengthVarNamesSubstring) {
+    if (lenFieldName.find(potentialName) != std::string::npos)
+      return true;
+  }
+  return false;
+}
+
+static bool hasLengthKeyword(std::string varName) {
+  // Convert the field name to lower case.
+  std::transform(varName.begin(), varName.end(), varName.begin(),
+                 [](unsigned char c){ return std::tolower(c); });
+
+  std::set<std::string> allLengthKeywords(possibleLengthVarNamesPrefixes);
+  allLengthKeywords.insert(possibleLengthVarNamesSubstring.begin(),
+                           possibleLengthVarNamesSubstring.end());
+  for (auto &potentialName : allLengthKeywords) {
+    if (varName.find(potentialName) != std::string::npos)
+      return true;
+  }
+  return false;
+}
+
+// Check if the provided constraint variable is an array and it needs bounds.
+static bool needArrayBounds(ConstraintVariable *CV,
+                            Constraints::EnvironmentMap &envMap) {
+  if (CV->hasArr(envMap)) {
+    PVConstraint *PV = dyn_cast<PVConstraint>(CV);
+    if (PV && PV->getArrPresent())
+      return false;
+    return true;
+  }
+  return false;
+}
+
+static bool needNTArrayBounds(ConstraintVariable *CV,
+                              Constraints::EnvironmentMap &envMap) {
+  if (CV->hasNtArr(envMap)) {
+    PVConstraint *PV = dyn_cast<PVConstraint>(CV);
+    if (PV && PV->getArrPresent())
+      return false;
+    return true;
+  }
+  return false;
+}
+
+static bool needArrayBounds(Decl *decl, ProgramInfo &Info, ASTContext *Context) {
```

Comment:
Please make the variable naming uniform throughout your changes. According to LLVM coding standards variable names should begin with an upper-case letter. See https://llvm.org/docs/CodingStandards.html#id42